### ISSUE TITLE
[YSQL] Batch Updates are inefficient when multiple writes land on same row

### ISSUE THREAD

**Issue by karthik-ramanathan-3006** [yugabyte/yugabyte-db#19939]
Jira Link: [DB-8906](https://yugabyte.atlassian.net/browse/DB-8906)
### Description

Create the following schema:

Consider a table with the following schema:
```
CREATE TABLE foo(h INT PRIMARY KEY, v1 INT, v2 INT);

-- Create an index on v1 covering v2
CREATE INDEX foo_v1_v2 ON foo (v1) INCLUDE (v2);

-- Insert a 1000 rows
INSERT INTO foo (SELECT i, i, i FROM generate_series(1, 1000) AS i);

yugabyte=# \d foo
                Table "public.foo"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 h      | integer |           | not null |
 v1     | integer |           |          |
 v2     | integer |           |          |
Indexes:
    "foo_pkey" PRIMARY KEY, lsm (h HASH)
    "foo_v1_v2" lsm (v1 HASH) INCLUDE (v2)
```

**Example 1**
UPDATE v2 across all rows.
```
yugabyte=# EXPLAIN (ANALYZE, DIST) UPDATE foo SET v2 = v2 + 1;
                                                 QUERY PLAN
------------------------------------------------------------------------------------------------------------
 Update on foo  (cost=0.00..102.50 rows=1000 width=80) (actual time=657.940..657.940 rows=0 loops=1)
   ->  Seq Scan on foo  (cost=0.00..102.50 rows=1000 width=80) (actual time=2.060..3.070 rows=1000 loops=1)
         Storage Table Read Requests: 2
         Storage Table Read Execution Time: 1.126 ms
         Storage Table Write Requests: 1000
         Storage Index Write Requests: 2000
         Storage Flush Requests: 1000                              <--- inefficient
         Storage Flush Execution Time: 271.284 ms
 Planning Time: 7.970 ms
 Execution Time: 661.673 ms
 Storage Read Requests: 2
 Storage Read Execution Time: 1.126 ms
 Storage Write Requests: 3000
 Catalog Read Requests: 7
 Catalog Read Execution Time: 15.023 ms
 Catalog Write Requests: 0
 Storage Flush Requests: 1001                                      <--- inefficient
 Storage Flush Execution Time: 271.520 ms
 Storage Execution Time: 287.669 ms
 Peak Memory Usage: 9 kB
(20 rows)
```
 
 There are two issues here:
 1. UPDATEs to the non-key columns in a covering index should not trigger a rewrite of the index row.
 2. (Focus of this issue) The updates/writes issued to the index have the pattern:
     -  DELETE (key = 1)
     -  INSERT (key = 1) 
     -  DELETE (key = 2) 
     -  INSERT (key = 2) 
     -  .... 
     - DELETE (key = 1000) 
     - INSERT (key = 1000)
     
Our current logic dictates that if there are two write operations to the same row, we perform a flush between the two operations. This causes 1000 + 1 flushes. The ideal behavior here would be to group the DELETEs together, flush and then perform the INSERTs, resulting in 2 flushes, rather than `n` flushes.

**Example 2**
```
yugabyte=# EXPLAIN (ANALYZE, DIST) UPDATE foo SET v1 = h;
                                                 QUERY PLAN
------------------------------------------------------------------------------------------------------------
 Update on foo  (cost=0.00..100.00 rows=1000 width=80) (actual time=646.538..646.538 rows=0 loops=1)
   ->  Seq Scan on foo  (cost=0.00..100.00 rows=1000 width=80) (actual time=9.005..9.855 rows=1000 loops=1)
         Storage Table Read Requests: 2
         Storage Table Read Execution Time: 7.240 ms
         Storage Table Write Requests: 1000
         Storage Index Write Requests: 2000
         Storage Flush Requests: 1000
         Storage Flush Execution Time: 258.264 ms
 Planning Time: 9.341 ms
 Execution Time: 647.049 ms
 Storage Read Requests: 2
 Storage Read Execution Time: 7.240 ms
 Storage Write Requests: 3000
 Catalog Read Requests: 3
 Catalog Read Execution Time: 7.811 ms
 Catalog Write Requests: 0
 Storage Flush Requests: 1001
 Storage Flush Execution Time: 258.499 ms
 Storage Execution Time: 273.551 ms
 Peak Memory Usage: 89 kB
(20 rows)
```

In this example, the value of `v1` is identical to `h` for all rows. #19305 will fix the issue of unmodified columns needing a rewrite/update but this example follows the same alternating DELETE + INSERT pattern described above.

To summarize:
When there are multiple pending write operations to the same row, it might be prudent to wait until the statement boundary (or some such marker), group pending operations by type and flush each distinct group in order of arrival.

### Issue Type

kind/enhancement

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-8906]: https://yugabyte.atlassian.net/browse/DB-8906?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by karthik-ramanathan-3006** [comment]
Covered by a combination of #18822 and #20908.
